controllers/utils/cert.go

```diff
@@ -0,0 +1,171 @@
+package utils
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	k8serr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func CreateSelfSignedCertificate(ctx context.Context, c client.Client, secretName, domain, namespace string) error {
+	certSecret, err := GenerateSelfSignedCertificateAsSecret(secretName, domain, namespace)
+	if err != nil {
+		return fmt.Errorf("failed generating self-signed certificate: %w", err)
+	}
+
+	if errGen := generateCertSecret(ctx, c, certSecret); errGen != nil {
+		return fmt.Errorf("failed update self-signed certificate secret: %w", errGen)
+	}
+
+	return nil
+}
+
+func GenerateSelfSignedCertificateAsSecret(name, addr, namespace string) (*corev1.Secret, error) {
+	cert, key, err := generateCertificate(addr)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+			Labels: map[string]string{
+				"opendatahub.io/managed":       "true",
+				"app.kubernetes.io/name":       "odh-model-controller",
+				"app.kubernetes.io/component":  "kserve",
+				"app.kubernetes.io/part-of":    "odh-model-serving",
+				"app.kubernetes.io/managed-by": "odh-model-controller",
+			},
+		},
+		Data: map[string][]byte{
+			corev1.TLSCertKey:       cert,
+			corev1.TLSPrivateKeyKey: key,
+		},
+		Type: corev1.SecretTypeTLS,
+	}, nil
+}
+
+func generateCertificate(addr string) ([]byte, []byte, error) {
+	key, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+
+	seededRand, cryptErr := rand.Int(rand.Reader, big.NewInt(time.Now().UnixNano()))
+	if cryptErr != nil {
+		return nil, nil, errors.WithStack(cryptErr)
+	}
+
+	now := time.Now()
+	tmpl := x509.Certificate{
+		SerialNumber: seededRand,
+		Subject: pkix.Name{
+			CommonName:   addr,
+			Organization: []string{"serving-self-signed"},
+		},
+		NotBefore:             now.UTC(),
+		NotAfter:              now.Add(time.Second * 60 * 60 * 24 * 365 * 10).UTC(),
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+	}
+
+	certDERBytes, err := x509.CreateCertificate(rand.Reader, &tmpl, &tmpl, key.Public(), key)
+	if err != nil {
+		return nil, nil, errors.WithStack(err)
```

Comment:
is there a specific use case for using `errors.WithStack` instead of `fmt.Errorf`
It can remove one dependency

Reply:
good catch! I will update it

Reply:
After looking at the code again, I changed my mind. I remembered why it used `errors.WithStack`. This is to force a stack when generating the certificate, so that if there is a problem, I can see where the problem is occurring. So I will not change it.